File: Extensification/Collections/Array/Getting.vb

'    Extensification  Copyright (C) 2020-2021  EoflaOE
'
'    This file is part of Extensification
'
'    Extensification is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Extensification is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Runtime.CompilerServices

Namespace ArrayExts
    ''' <summary>
    ''' Provides the array extensions related to getting
    ''' </summary>
    Public Module Getting

        ''' <summary>
        ''' Gets index from entry
        ''' </summary>
        ''' <typeparam name="T">Type</typeparam>
        ''' <param name="TargetArray">Target array</param>
        ''' <param name="Entry">An entry from array</param>
        ''' <returns>List of indexes. If none is found, returns an empty array list</returns>
        <Extension>
        Public Function GetIndexFromEntry(Of T)(TargetArray() As T, Entry As T) As Integer()
            If TargetArray Is Nothing Then Throw New ArgumentNullException(NameOf(TargetArray))
            Dim Indexes As New ArrayList
            For Index As Integer = 0 To TargetArray.Length - 1
                Dim ArrayEntry As Object = TargetArray(Index)
                If ArrayEntry = Entry Then
                    Indexes.Add(Index)
                End If
            Next
            Return Indexes.ToArray(GetType(Integer))
        End Function

        ''' <summary>
        ''' Gets indexes of non-empty items
        ''' </summary>
        ''' <typeparam name="T">Type</typeparam>
        ''' <param name="TargetArray">Target array</param>
        ''' <returns>Indexes of non-empty items</returns>
        <Extension>
        Public Function GetIndexesOfFullEntries(Of T)(TargetArray() As T) As Integer()
            Dim FullIndexes As New List(Of Integer)
            For i As Long = 0 To TargetArray.LongLength - 1
                If TargetArray(i) IsNot Nothing Then
                    If TypeOf TargetArray(i) Is String Then
                        If Not TargetArray(i).Equals("") Then
                            FullIndexes.Add(i)
                        End If
                    Else
                        FullIndexes.Add(i)
                    End If
                End If
            Next
            Return FullIndexes.ToArray
        End Function

        ''' <summary>
        ''' Gets indexes of empty items
        ''' </summary>
        ''' <typeparam name="T">Type</typeparam>
        ''' <param name="TargetArray">Target array</param>
        ''' <returns>Indexes of empty items</returns>
        <Extension>
        Public Function GetIndexesOfEmptyEntries(Of T)(TargetArray() As T) As Integer()
            Dim EmptyIndexes As New List(Of Integer)
            For i As Long = 0 To TargetArray.LongLength - 1
                If TargetArray(i) Is Nothing Then
                    EmptyIndexes.Add(i)
                ElseIf TypeOf TargetArray(i) Is String And TargetArray(i).Equals("") Then
                    EmptyIndexes.Add(i)
                End If
            Next
            Return EmptyIndexes.ToArray
        End Function

    End Module
End Namespace

File: Extensification/Collections/Array/Getting.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Microsoft.VisualBasic.CompilerServices;

namespace Extensification.ArrayExts
{
    /// <summary>
    /// Provides the array extensions related to getting
    /// </summary>
    public static class Getting
    {

        /// <summary>
        /// Gets index from entry
        /// </summary>
        /// <typeparam name="T">Type</typeparam>
        /// <param name="TargetArray">Target array</param>
        /// <param name="Entry">An entry from array</param>
        /// <returns>List of indexes. If none is found, returns an empty array list</returns>
        public static int[] GetIndexFromEntry<T>(this T[] TargetArray, T Entry)
        {
            if (TargetArray is null)
                throw new ArgumentNullException(nameof(TargetArray));
            var Indexes = new ArrayList();
            for (int Index = 0, loopTo = TargetArray.Length - 1; Index <= loopTo; Index++)
            {
                object ArrayEntry = TargetArray[Index];
                if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(ArrayEntry, Entry, false)))
                {
                    Indexes.Add(Index);
                }
            }
            return (int[])Indexes.ToArray(typeof(int));
        }

        /// <summary>
        /// Gets indexes of non-empty items
        /// </summary>
        /// <typeparam name="T">Type</typeparam>
        /// <param name="TargetArray">Target array</param>
        /// <returns>Indexes of non-empty items</returns>
        public static int[] GetIndexesOfFullEntries<T>(this T[] TargetArray)
        {
            var FullIndexes = new List<int>();
            for (long i = 0L, loopTo = TargetArray.LongLength - 1L; i <= loopTo; i++)
            {
                if (TargetArray[(int)i] is not null)
                {
                    if (Conversions.ToBoolean(TargetArray[(int)i] is string))
                    {
                        if (!TargetArray[(int)i].Equals(""))
                        {
                            FullIndexes.Add((int)i);
                        }
                    }
                    else
                    {
                        FullIndexes.Add((int)i);
                    }
                }
            }
            return FullIndexes.ToArray();
        }

        /// <summary>
        /// Gets indexes of empty items
        /// </summary>
        /// <typeparam name="T">Type</typeparam>
        /// <param name="TargetArray">Target array</param>
        /// <returns>Indexes of empty items</returns>
        public static int[] GetIndexesOfEmptyEntries<T>(this T[] TargetArray)
        {
            var EmptyIndexes = new List<int>();
            for (long i = 0L, loopTo = TargetArray.LongLength - 1L; i <= loopTo; i++)
            {
                if (TargetArray[(int)i] is null)
                {
                    EmptyIndexes.Add((int)i);
                }
                else if (Conversions.ToBoolean(TargetArray[(int)i] is string) & TargetArray[(int)i].Equals(""))
                {
                    EmptyIndexes.Add((int)i);
                }
            }
            return EmptyIndexes.ToArray();
        }

    }
}